filebeat/input/file/state.go

```diff
@@ -81,25 +83,36 @@ func (s *States) findPrevious(newState State) (int, State) {
 }
 
 // Cleanup cleans up the state array. All states which are older then `older` are removed
-func (s *States) Cleanup(older time.Duration) {
+func (s *States) Cleanup() {
+
 	s.mutex.Lock()
 	defer s.mutex.Unlock()
 
 	for i, state := range s.states {
 
-		// File wasn't seen for longer then older -> remove state
-		if time.Since(state.LastSeen) > older {
-			logp.Debug("prospector", "State removed for %s because of older: %s", state.Source)
-			s.states = append(s.states[:i], s.states[i+1:]...)
+		// -1 means ttl is disabled
+		if state.Ttl == -1 {
+			continue
 		}
-	}
 
+		// States is older then ttl
+		if int64(time.Since(state.Timestamp).Seconds()) > state.Ttl {
+			// This check prevents a panic
+			if len(s.states) == 1 {
+				s.states = []State{}
+			} else {
+				s.states = append(s.states[:i], s.states[i+1:]...)
+			}
```

Comment:
The registrar does the cleanup himself. As all states are reported to prospector and registrar, they will have the same information -> do the same cleanup. The cleanup on the registrar side will always happen with a delay. It can happen in an edge case, that cleanup removes a state on the registrar side that was not removed on the prospector side because a new event already exist. But the new state will be sent to the registrar and reintroduce the state again. Prospector always overwrites registrar state.

We should discuss in a second step on the requirements for variables as some combinations / variables can lead to very strange situations. If clean_older is too small, filebeat will "stateless".